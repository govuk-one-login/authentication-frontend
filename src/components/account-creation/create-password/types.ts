import type { ApiResponseResult, DefaultApiResponse } from "../../../types.js";
import type { Request } from "express";

export interface CreatePasswordServiceInterface {
  signUpUser: (
    sessionId: string,
    clientSessionId: string,
    emailAddress: string,
    password: string,
    persistentSessionId: string,
    req: Request
  ) => Promise<ApiResponseResult<DefaultApiResponse>>;
}
