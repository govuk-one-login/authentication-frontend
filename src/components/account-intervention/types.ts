import { ApiResponseResult, DefaultApiResponse } from "../../types";

export interface AccountInterventionsInterface {
  accountInterventionStatus: (
    sessionId: string,
    emailAddress: string,
    sourceIp: string,
    clientSessionId: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<AccountInterventionStatus>>;
}

export interface AccountInterventionStatus extends DefaultApiResponse {
  email: string;
  passwordResetRequired: boolean;
  blocked: boolean;
  temporarilySuspended: boolean;
  appliedAt: string;
  reproveIdentity: boolean;
}
