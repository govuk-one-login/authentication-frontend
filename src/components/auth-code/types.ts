import {
  ApiResponseResult,
  DefaultApiResponse,
  UserSession,
  UserSessionClient,
} from "../../types";
import { Request } from "express";

export interface AuthCodeResponse extends DefaultApiResponse {
  location: string;
}

export interface AuthCodeServiceInterface {
  getAuthCode: (
    sessionId: string,
    clientSessionId: string,
    sourceIp: string,
    persistentSessionId: string,
    clientSession: UserSessionClient,
    userSession: UserSession,
    req: Request
  ) => Promise<ApiResponseResult<AuthCodeResponse>>;
}
