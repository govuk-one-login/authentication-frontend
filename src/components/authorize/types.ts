import type { MFA_METHOD_TYPE } from "../../app.constants.js";
import type { ApiResponseResult, DefaultApiResponse } from "../../types.js";
import type { Claims } from "./claims-config.js";
import type { Request } from "express";

export interface StartRequestParameters {
  authenticated: boolean;
  previous_session_id?: string;
  rp_pairwise_id_for_reauth?: string;
  previous_govuk_signin_journey_id?: string;
  reauthenticate?: string;
  cookie_consent?: string;
  _ga?: string;
  rp_client_id: string;
  scope: string;
  rp_redirect_uri: string;
  rp_state: string;
  requested_level_of_confidence?: string;
  requested_credential_strength: string;
  client_name: string;
  service_type: string;
  cookie_consent_shared: boolean;
  is_smoke_test: boolean;
  is_one_login_service: boolean;
  subject_type: string;
  is_identity_verification_required: boolean;
  rp_sector_host: string;
}

export interface StartAuthResponse extends DefaultApiResponse {
  user: UserSessionInfo;
  featureFlags?: Record<string, unknown>;
}

export interface UserSessionInfo {
  upliftRequired: boolean;
  identityRequired: boolean;
  authenticated: boolean;
  cookieConsent?: string;
  gaCrossDomainTrackingId?: string;
  mfaMethodType?: MFA_METHOD_TYPE;
  isBlockedForReauth: boolean;
}

export interface AuthorizeServiceInterface {
  start: (
    sessionId: string,
    clientSessionId: string,
    persistentSessionId: string,
    req: Request,
    startRequestParameters: StartRequestParameters
  ) => Promise<ApiResponseResult<StartAuthResponse>>;
}

export interface KmsDecryptionServiceInterface {
  decrypt(serializedJwe: string): Promise<string>;
}

export interface JwtServiceInterface {
  getPayloadWithValidation(jwt: string): Promise<Claims>;
}
