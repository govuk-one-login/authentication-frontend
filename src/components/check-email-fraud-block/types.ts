import type { ApiResponseResult, DefaultApiResponse } from "../../types.js";
import type { Request } from "express";

export interface CheckEmailFraudBlockInterface {
  checkEmailFraudBlock: (
    email: string,
    sessionId: string,
    clientSessionId: string,
    persistentSessionId: string,
    req: Request
  ) => Promise<ApiResponseResult<CheckEmailFraudBlockResponse>>;
}

export interface CheckEmailFraudBlockResponse extends DefaultApiResponse {
  email: string;
  isBlockedStatus: string;
}
