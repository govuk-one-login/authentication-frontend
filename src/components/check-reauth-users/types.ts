import { ApiResponseResult, DefaultApiResponse } from "../../types";

export interface CheckReauthServiceInterface {
  checkReauthUsers: (
    sessionId: string,
    email: string,
    sourceIp: string,
    clientSessionId: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<DefaultApiResponse>>;
}
