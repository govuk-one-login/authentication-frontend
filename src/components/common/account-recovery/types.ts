import { ApiResponseResult, DefaultApiResponse } from "../../../types";

export interface AccountRecoveryResponse extends DefaultApiResponse {
  accountRecoveryPermitted: boolean;
}

export interface AccountRecoveryInterface {
  accountRecovery: (
    sessionId: string,
    clientSessionId: string,
    email: string,
    sourceIp: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<AccountRecoveryResponse>>;
}
