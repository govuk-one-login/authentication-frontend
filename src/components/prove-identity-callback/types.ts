import { ApiResponseResult, DefaultApiResponse } from "../../types";

export interface ProveIdentityCallbackServiceInterface {
  processIdentity: (
    email: string,
    sourceIp: string,
    sessionId: string,
    clientSessionId: string,
    persistentSessionId: string
  ) => Promise<ApiResponseResult<ProcessIdentityResponse>>;
}

export type ProcessIdentityResponse =
  | ProcessIdentitySPOTResponse
  | ProcessIdentityInterventionResponse;

interface ProcessIdentitySPOTResponse extends DefaultApiResponse {
  status:
    | IdentityProcessingStatus.COMPLETED
    | IdentityProcessingStatus.PROCESSING
    | IdentityProcessingStatus.ERROR;
}

interface ProcessIdentityInterventionResponse extends DefaultApiResponse {
  status: IdentityProcessingStatus.INTERVENTION;
  redirectUrl: string;
}

export enum IdentityProcessingStatus {
  COMPLETED = "COMPLETED",
  ERROR = "ERROR",
  PROCESSING = "PROCESSING",
  INTERVENTION = "INTERVENTION",
}
