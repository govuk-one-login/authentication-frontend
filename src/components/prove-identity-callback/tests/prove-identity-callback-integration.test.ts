import { PATH_NAMES } from "../../../app.constants.js";
import { describe } from "mocha";
import { request, sinon } from "../../../../test/utils/test-utils.js";
import type { NextFunction, Request, Response } from "express";
import type express from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper.js";
import decache from "decache";
import { IdentityProcessingStatus } from "../types.js";
describe("Integration: prove identity callback", () => {
  let app: express.Application;

  describe("permission to visit /ipv-callback", () => {
    it(`should allow visit to /ipv-callback when was on /auth-code`, async () => {
      // arrange
      const redirectPath = "https://test.example.com";
      app = await stubMiddlewareAndCreateApp(
        PATH_NAMES.AUTH_CODE,
        redirectPath
      );

      // act
      await request(app, (test) =>
        test
          .get(PATH_NAMES.PROVE_IDENTITY_CALLBACK)
          // assert
          .expect(302)
          .expect("Location", redirectPath)
      );
    });

    it("should redirect when not on permitted journey to /ipv-callback", async () => {
      // arrange
      const startingPage = PATH_NAMES.ROOT;
      app = await stubMiddlewareAndCreateApp(startingPage);

      // act
      await request(app, (test) =>
        test
          .get(PATH_NAMES.PROVE_IDENTITY_CALLBACK)
          // assert
          .expect(302)
          .expect("Location", startingPage)
      );
    });
  });
});

const stubMiddlewareAndCreateApp = async (
  previousPath: string,
  redirectPath?: string
) => {
  decache("../../../app");

  decache("../../../middleware/session-middleware");
  const sessionMiddleware = await import(
    "../../../middleware/session-middleware.js"
  );
  sinon
    .stub(sessionMiddleware, "validateSessionMiddleware")
    .callsFake(function (
      req: Request,
      res: Response,
      next: NextFunction
    ): void {
      res.locals.sessionId = "testSessionId";
      res.locals.clientSessionId = "testClientSessionId";
      res.locals.persistentSessionId = "testPersistentSessionId";

      req.session.client = {
        name: "testClientName",
      };
      req.session.user = {
        email: "test@test.com",
        journey: getPermittedJourneyForPath(previousPath),
      };

      next();
    });

  if (redirectPath) {
    decache("../prove-identity-callback-service");
    const proveIdentityCallbackServiceFile = await import(
      "../prove-identity-callback-service.js"
    );
    sinon
      .stub(proveIdentityCallbackServiceFile, "proveIdentityCallbackService")
      .returns({
        processIdentity: sinon.fake.resolves({
          data: { status: IdentityProcessingStatus.COMPLETED },
        }),
        generateSuccessfulRpReturnUrl: sinon.fake.resolves(redirectPath),
      });
  }

  return (await import("../../../app.js")).createApp();
};
