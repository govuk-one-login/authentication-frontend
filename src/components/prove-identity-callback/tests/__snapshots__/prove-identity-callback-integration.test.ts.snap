// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: prove identity callback permission to visit /ipv-callback should redirect when not on permitted journey to /ipv-callback 1`] = `
Object {
  "contentId": "3104ec55-1a4e-4811-b927-0531fb315480",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
